crates/luminol-core/src/window.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright (C) 2023 Lily Lyons
+//
+// This file is part of Luminol.
+//
+// Luminol is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Luminol is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Luminol.  If not, see <http://www.gnu.org/licenses/>.
+//
+//     Additional permission under GNU GPL version 3 section 7
+//
+// If you modify this Program, or any covered work, by linking or combining
+// it with Steamworks API by Valve Corporation, containing parts covered by
+// terms of the Steamworks API by Valve Corporation, the licensors of this
+// Program grant you additional permission to convey the resulting work.
+
+/// A window management system to handle heap allocated windows
+///
+/// Will deny any duplicated window titles and is not specialized like modals
+#[derive(Default)]
+pub struct Windows {
+    // A dynamic array of Windows. Iterated over and cleaned up in fn update().
+    windows: Vec<Box<dyn Window>>,
+}
+
+#[derive(Default)]
+pub struct EditWindows {
+    clean_fn: Option<CleanFn>,
+    added: Vec<Box<dyn Window>>,
+    removed: std::collections::HashSet<egui::Id>,
+}
+
+type CleanFn = Box<dyn Fn(&Box<dyn Window>) -> bool>;
+
+impl Windows {
+    /// A function to add a window.
+    pub fn add_window(&mut self, window: impl Window + 'static) {
+        self.add_boxed_window(Box::new(window))
+    }
+
+    fn add_boxed_window(&mut self, window: Box<dyn Window>) {
+        // FIXME use a hashmap, or something with less than O(n) search time
+        if self.windows.iter().any(|w| w.id() == window.id()) {
+            return;
+        }
+        self.windows.push(window)
+    }
```

Comment:
Who cares if this is O(n)? There's no way the user is going to add more than one window or tab per frame.

If it really matters that much, you can compute a hashmap of IDs at the beginning of the `display` method, which doesn't lose performance since we need to iterate over the windows once per frame just to draw them.

Reply:
Good point